Fix ronja writes eagerness to work with unwind

henriknyman opened the pull request.

henriknyman commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:
```diff
@@ -1362,21 +1365,49 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
+    val result = updateWithBothPlanners(query)
+    result.columnAs[Int]("count").next should equal(14)
+    assertStats(result, propertiesSet = 14, nodesCreated = 3)
+    assertNumberOfEagerness(query, 3)
```
I think this should just need 2 eagerness if we were smarter with set property overlap.. (the top repeatable read should be unnecessary)

henriknyman commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/PlannerQuery.scala:
```diff
@@ -188,6 +188,100 @@ object PlannerQuery {
+    def conflicts(plannerQuery: PlannerQuery, next: PlannerQuery): Boolean = {
+      val conflict = if (next.updateGraph.isEmpty) false
+      else if (!fistLeafIsStable)
+        next.updateGraph overlaps plannerQuery.queryGraph
```
We could refactor this to be a bit more pretty. Now it got a bit smashed together

henriknyman commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/PlannerQuery.scala:
```diff
@@ -188,6 +188,100 @@ object PlannerQuery {
+        (None, leaves)
+
+    @tailrec
+    def conflicts(plannerQuery: PlannerQuery, next: PlannerQuery): Boolean = {
```
Better parameter names? Or if we move the method to the PlannerQuery class then `plannerQuery` will just be `this`

henriknyman commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/PlannerQuery.scala:
```diff
@@ -188,6 +188,100 @@ object PlannerQuery {
+        next.updateGraph overlaps plannerQuery.queryGraph
+      else {
+        //if we have unsafe rels we need to check relation overlap and delete
+        //overlap immediately
```
Maybe this could be broken out as a separate special case `overlaps` method on updateGraph for the first planner query?

Mats-SX commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:
```diff
@@ -1439,10 +1470,21 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
+  test("matching label on left-hand side and removing same label on right-hand side should not be eager") {
+    createLabeledNode("Lol")
+    createNode()
+    val query = "MATCH (m:Lol), (n) REMOVE n:Lol RETURN count(*)"
```
There's no guarantee here that the label scan goes to the left with the cost planner.

Mats-SX commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:
```diff
@@ -1439,10 +1470,21 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
     createLabeledNode("Lol")
     createNode()
-    val query = "MATCH  (m:Lol), (n) REMOVE n:Lol RETURN count(*)"
+    val query = "MATCH (m:Lol), (n) REMOVE n:Lol RETURN count(*)"
```
Likewise there's no guarantee it goes to the right either...

Mats-SX committed:
Normalize eagerization testing

- All eagerization tests now assert on how many rows are returned.
- Moved a test that was not eager-specific out of the file
- `updateWithBothPlanners` will now check eager-correctness before rule/cost

Mats-SX committed:
Restructure how eagerness analysis is performed

- Fixes a bug with `REMOVE` where the same label is read multiple times
- Simplifies the functions used to do eagerness analysis

Mats-SX commented:
This PR might fix the problem stated in the title, but is missing a number of acceptance tests with queries using `UNWIND`.

Mats-SX committed:
Add extra acceptance tests with `UNWIND`

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/Eagerness.scala:
```diff
@@ -0,0 +1,134 @@
+
+  @tailrec
+  private def headConflicts(head: PlannerQuery, tail: PlannerQuery, unstableLeaves: Seq[IdName]): Boolean = {
+    val conflict = if (tail.updateGraph.isEmpty) false
```
This should not stop the recursion, but continue to the next tail.

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/Eagerness.scala:
```diff
@@ -0,0 +1,134 @@
+  def conflictInHead(plan: LogicalPlan, plannerQuery: PlannerQuery): Boolean = {
+    // The first leaf node is always reading through a stable iterator.
+    // We will only consider this analysis for all other node iterators.
+    val unstableLeaves = plan.leaves.collect {
```
Should be called `leaves`.

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/Eagerness.scala:
```diff
@@ -0,0 +1,134 @@
+    if (unstableLeaves.isEmpty)
+      false // the query did not start with a read, possibly CREATE () ...
+    else
+      // Start recursion by checking the given plannerQuery against itself
```
Add comment about stable leaf.

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/Eagerness.scala:
```diff
@@ -0,0 +1,134 @@
+        //MATCH (n:A), (m:B) REMOVE n:B
+        //MATCH (n:A), (m:A) REMOVE m:A
+        (labelsToRemove intersect labelsOnCurrentNode).nonEmpty
+        )
```
Silly linebreak.

Mats-SX committed:
WIP -- fixes a few bugs with eagerness

- remove labels
- delete/merge combinations

EXPLORE: put merges inside own PlannerQuery -- divide PlannerQuery into inheritance structure with normal PlannerQuery and MergePlannerQuery

systay commented:
Re-open when this should be merged

Mats-SX commented:
This PR can just die.
